File: ios/Widgets/Sources/TasbihWidget.swift
import AppIntents
import CoreKit
import SwiftUI
import WidgetKit

private let appGroup = "group.com.fatwabot.app"

private func counterStore() -> TasbihWidgetCounterStore? {
    FileManager.default
        .containerURL(forSecurityApplicationGroupIdentifier: appGroup)
        .map { TasbihWidgetCounterStore(appGroupContainer: $0) }
}

/// Adds one to the tally. Kept separate from the reset intent rather than
/// parameterised, because the two carry very different weight: increment is the
/// whole point of the widget and reset destroys a count someone has been
/// accumulating. Distinct intents make the destructive one impossible to invoke
/// by passing the wrong value.
struct IncrementTasbihIntent: AppIntent {
    static var title: LocalizedStringResource = "Count"
    static var openAppWhenRun: Bool = false

    init() {}

    func perform() async throws -> some IntentResult {
        if let store = counterStore() {
            store.write(store.read().incremented())
        }
        WidgetCenter.shared.reloadTimelines(ofKind: "TasbihWidget")
        return .result()
    }
}

struct ResetTasbihIntent: AppIntent {
    static var title: LocalizedStringResource = "Reset count"
    static var openAppWhenRun: Bool = false

    init() {}

    func perform() async throws -> some IntentResult {
        if let store = counterStore() {
            store.write(store.read().reset())
        }
        WidgetCenter.shared.reloadTimelines(ofKind: "TasbihWidget")
        return .result()
    }
}

// MARK: - Provider

struct TasbihEntry: TimelineEntry {
    let date: Date
    let count: Int
}

struct TasbihProvider: TimelineProvider {
    func placeholder(in context: Context) -> TasbihEntry {
        TasbihEntry(date: Date(), count: 0)
    }

    func getSnapshot(in context: Context, completion: @escaping (TasbihEntry) -> Void) {
        let now = Date()
        completion(TasbihEntry(date: now, count: counterStore()?.read().current(on: now) ?? 0))
    }

    func getTimeline(in context: Context, completion: @escaping (Timeline<TasbihEntry>) -> Void) {
        let now = Date()
        let entry = TasbihEntry(date: now, count: counterStore()?.read().current(on: now) ?? 0)
        // Refresh at the start of tomorrow so the tally visibly rolls over
        // rather than waiting for the next tap to reveal it had reset.
        let tomorrow = Calendar.current.nextDate(
            after: now, matching: DateComponents(hour: 0, minute: 0), matchingPolicy: .nextTime
        ) ?? now.addingTimeInterval(3600)
        completion(Timeline(entries: [entry], policy: .after(tomorrow)))
    }
}

// MARK: - Widget

struct TasbihWidget: Widget {
    var body: some WidgetConfiguration {
        StaticConfiguration(kind: "TasbihWidget", provider: TasbihProvider()) { entry in
            TasbihWidgetView(entry: entry)
                .brandWidgetContainer()
        }
        .configurationDisplayName(Text("widget.tasbih.name"))
        .description(Text("widget.tasbih.desc"))
        .supportedFamilies([.systemSmall, .systemMedium])
    }
}

struct TasbihWidgetView: View {
    let entry: TasbihEntry

    var body: some View {
        ZStack(alignment: .topLeading) {
            // The whole tile counts. A small centre button would be the obvious
            // layout and the wrong one — dhikr is repeated without looking, so
            // the target should be everything the thumb can land on.
            Button(intent: IncrementTasbihIntent()) {
                VStack(spacing: 2) {
                    Text("\(entry.count)")
                        .font(.system(size: 44, weight: .bold, design: .rounded))
                        .monospacedDigit()
                        .foregroundStyle(brandPrimary)
                        .minimumScaleFactor(0.5)
                        .lineLimit(1)
                    Text("widget.tasbih.hint")
                        .font(.caption2)
                        .foregroundStyle(brandMuted)
                }
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(Text("widget.tasbih.name"))
            .accessibilityValue(Text("\(entry.count)"))

            // Reset sits in a corner, deliberately small: it must be reachable
            // without being in the path of a thumb tapping the same tile a
            // hundred times.
            Button(intent: ResetTasbihIntent()) {
                Image(systemName: "arrow.counterclockwise")
                    .font(.caption2)
                    .foregroundStyle(brandMuted)
                    .padding(6)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(Text("widget.tasbih.reset"))
        }
    }
}

File: android/widget/src/main/kotlin/com/fatwabot/widget/TasbihWidget.kt
package com.fatwabot.widget

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.glance.GlanceId
import androidx.glance.GlanceModifier
import androidx.glance.GlanceTheme
import androidx.glance.action.ActionParameters
import androidx.glance.action.clickable
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.GlanceAppWidgetReceiver
import androidx.glance.appwidget.action.ActionCallback
import androidx.glance.appwidget.action.actionRunCallback
import androidx.glance.appwidget.provideContent
import androidx.glance.appwidget.updateAll
import androidx.glance.layout.Alignment
import androidx.glance.layout.Box
import androidx.glance.layout.Column
import androidx.glance.layout.fillMaxSize
import androidx.glance.layout.padding
import androidx.glance.text.FontWeight
import androidx.glance.text.Text
import androidx.glance.text.TextAlign
import androidx.glance.text.TextStyle
import com.fatwabot.core.common.TasbihWidgetCounter

/**
 * العداد — the interactive dhikr counter. Mirror of iOS `TasbihWidget`.
 *
 * The whole tile counts. A small centre button would be the obvious layout and
 * the wrong one: dhikr is repeated without looking, so the target should be
 * everything the thumb can land on. Reset sits in a corner, deliberately small,
 * so it is reachable without being in the path of a thumb tapping the same tile
 * a hundred times.
 */
class TasbihWidget : GlanceAppWidget() {
    override suspend fun provideGlance(context: Context, id: GlanceId) {
        val count = TasbihWidgetCounter.store(context.filesDir).read().current()
        provideContent {
            GlanceTheme { TasbihContent(context, count) }
        }
    }
}

class IncrementTasbihAction : ActionCallback {
    override suspend fun onAction(context: Context, glanceId: GlanceId, parameters: ActionParameters) {
        val store = TasbihWidgetCounter.store(context.filesDir)
        store.write(store.read().incremented())
        TasbihWidget().updateAll(context)
    }
}

class ResetTasbihAction : ActionCallback {
    override suspend fun onAction(context: Context, glanceId: GlanceId, parameters: ActionParameters) {
        val store = TasbihWidgetCounter.store(context.filesDir)
        store.write(store.read().reset())
        TasbihWidget().updateAll(context)
    }
}

@Composable
private fun TasbihContent(context: Context, count: Int) {
    Box(
        modifier = GlanceModifier
            .fillMaxSize()
            .brandSurface()
            .clickable(actionRunCallback<IncrementTasbihAction>()),
        contentAlignment = Alignment.Center,
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text(
                "$count",
                style = TextStyle(
                    color = MaroonProvider,
                    fontSize = 40.sp,
                    fontWeight = FontWeight.Bold,
                    textAlign = TextAlign.Center,
                ),
            )
            Text(
                context.getString(R.string.widget_tasbih_hint),
                style = TextStyle(color = MutedProvider, fontSize = 11.sp),
            )
        }
        Box(
            modifier = GlanceModifier
                .padding(6.dp)
                .clickable(actionRunCallback<ResetTasbihAction>()),
            contentAlignment = Alignment.TopStart,
        ) {
            Text(
                context.getString(R.string.widget_tasbih_reset),
                style = TextStyle(color = MutedProvider, fontSize = 11.sp),
            )
        }
    }
}

class TasbihWidgetReceiver : GlanceAppWidgetReceiver() {
    override val glanceAppWidget: GlanceAppWidget = TasbihWidget()
}
